go/parquet/internal/encoding/boolean_decoder.go

```diff
@@ -109,3 +114,76 @@ func (dec *PlainBooleanDecoder) DecodeSpaced(out []bool, nullCount int, validBit
 	}
 	return dec.Decode(out)
 }
+
+type RleBooleanDecoder struct {
+	decoder
+
+	rleDec *utils.RleDecoder
+}
+
+func (RleBooleanDecoder) Type() parquet.Type {
+	return parquet.Types.Boolean
+}
+
+func (dec *RleBooleanDecoder) SetData(nvals int, data []byte) error {
+	dec.nvals = nvals
+
+	if len(data) < 4 {
+		return fmt.Errorf("invalid length - %d (corrupt data page?)", len(data))
+	}
+
+	// load the first 4 bytes in little-endian which indicates the length
+	nbytes := binary.LittleEndian.Uint32(data[:4])
+	if nbytes > uint32(len(data)-4) {
+		return fmt.Errorf("received invalid number of bytes - %d (corrupt data page?)", nbytes)
+	}
+
+	dec.data = data[4:]
+	if dec.rleDec == nil {
+		dec.rleDec = utils.NewRleDecoder(bytes.NewReader(dec.data), 1)
+	} else {
+		dec.rleDec.Reset(bytes.NewReader(dec.data), 1)
+	}
+	return nil
+}
+
+func (dec *RleBooleanDecoder) Decode(out []bool) (int, error) {
+	max := shared_utils.MinInt(len(out), dec.nvals)
+
+	var (
+		buf [1024]uint64
+		n   = max
+	)
+
+	for n > 0 {
+		batch := shared_utils.MinInt(len(buf), n)
+		decoded := dec.rleDec.GetBatch(buf[:batch])
+		if decoded != batch {
+			return max - n, io.ErrUnexpectedEOF
```

Comment:
Is there a point in returning `max - n` here? What is the caller supposed to do with it?
(note you aren't actually writing out the decoded values, so I assume the decoded bytes are lost, which means the decoder can't be used anymore afterwards)

Reply:
We *are* writing out the decoded values via the loop at line 165. The input to this function is an output slice to write to and we populate it after the call to `GetBatch`. So returning `max - n` here informs the user of how many values were populated into that slice before the error was hit.

Reply:
But `dec.rleDec.GetBatch` has consumed some input and decoded some bytes that are just lost in `buf`, right?

Reply:
Correct. After receiving an error, a given decoder should not continue to be used, but we return the number of successfully output values so that a caller knows what values are there before the error was encountered.

Reply:
> We are writing out the decoded values via the loop at line 165. The input to this function is an output slice to write to and we populate it after the call to GetBatch. So returning max - n here informs the caller of how many values were populated into that slice before the error was hit.

No, `max` is not the real maximum of the underlying data. For rep-def existing scenerio, it means "rep-def" size, rather than the actual size. So it may greater than existing value.

Reply:
> No, max is not the real maximum of the underlying data. For rep-def existing scenerio, it means "rep-def" size, rather than the actual size. So it may greater than existing value. (At least this is the case in C++)

this is a level above where we are here. This is *solely* the value decoder. In this function `max` is the maximum number of *values* that are decoded by a single call to this function (i.e. the minimum of the length of the output slice and the number of values left to decode). 

The handling of rep and def levels is taken care of a level above this by the actual page reader / column chunk reader, which determines how many physical values it wants to read from the decoder. The intent of this function is that the return value is the *number of physical values populated into the output slice*.

Reply:
```go
func (dec *RleBooleanDecoder) SetData(nvals int, data []byte) error {
	dec.nvals = nvals


func (dec *RleBooleanDecoder) Decode(out []bool) (int, error) {
	max := shared_utils.MinInt(len(out), dec.nvals)
```


I totally agree your response. What I mean is that, during `SetData`, the `nvals` is `max-levels`. So in `Decode`,  `shared_utils.MinInt(len(out), dec.nvals)` would be more than existing value.

Reply:
.i.e

For:

```
nonnull: [false, false, false, false]
```

Decode(5) with length 10 will get `4, nil`

For 

```
nullable: [false, false, false, false, null]
```

Decode(5) will get `4, err`

Would this happens?

Reply:
No, that wouldn't happen because parquet doesn't include nulls in the underlying data. If the data is 

```
nullable: [false, false, false, false, null]`
```

Then the *physical data in the file* is just `[false, false, false, false]` and the def levels would have `[1, 1, 1, 1, 0]`.

Decode(5) would still get `4, nil` because only 4 *physical values* were decoded. This scenario is also tested in the unit tests.

Reply:
I know your meaning, but the problem is that, when we get a page. The `num_values` is including the nulls. ( see https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L533  ). So when:

```
nullable: [false, false, false, false, null]
```

The data size is `4`. But the `dec.nvals` is 5.

Reply:
This wouldn't matter without using raw-decode api. Because RecordReader will handing `Decode` well. But it's still make the syntax a bit inconsistent

Reply:
Hmm, you have a point though I think that would be a greater issue to think about attempting to fix. 

For DataPageV2Header it contains the number of nulls so we can easily handle changing `dec.nvals` to the correct number, but for DataPageV1 you have a point that *technically* this might be slightly off or is at least a bug waiting to happen that should be addressed.

Reply:
Yeah, lets create an issue for that. I think user would not easily touch it because we always suggest using RecordReader, right?

Reply:
or going through the ColumnChunkReader which has similar correct handling. We don't actually expose the raw decoder api for users to access at all.

Reply:
Aha that's right.